File: TemaProiect/motoProject/src/main/java/org/example/motoproject/repos/RaceRepo.java
package org.example.motoproject.repos;

import org.example.motoproject.domain.Participant;
import org.example.motoproject.domain.Race;

import java.sql.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

public class RaceRepo implements RaceRepoInterface {
    private final String url;
    private static final Logger logger = LogManager.getLogger(RaceRepo.class);

    RaceRepo(String url) {
        this.url = url;
    }

    @Override
    public void save(Race race) {
        String query = "INSERT INTO races (name, engine_size, date) VALUES (?, ?, ?)";
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query)) {
            logger.info("Saving race: {}", race);
            stmt.setString(1, race.getName());
            stmt.setLong(2, race.getEngineSize());
            stmt.setString(3, race.getDate());
            stmt.executeUpdate();
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }
    }

    @Override
    public Optional<Race> findById(Long id) {
        String query1 = "SELECT * FROM races WHERE id = ?";
        String query2 = "SELECT * FROM races_participants WHERE race_id = ?";

        Long raceId = null, engineSize = null;
        String name = null, date = null;
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query1)) {
            logger.info("Finding race by id: {}", id);
            stmt.setLong(1, id);
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                raceId = rs.getLong("id");
                name = rs.getString("name");
                engineSize = rs.getLong("engine_size");
                date = rs.getString("date");
            }
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }

        if (raceId == null) {
            logger.info("Race not found: {}", id);
            return Optional.empty();
        }

        List<Long> participantIds = new ArrayList<>();
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query2)) {
            logger.info("Finding participants by race id: {}", id);
            stmt.setLong(1, id);
            ResultSet rs = stmt.executeQuery();
            while (rs.next()) {
                Long participantId = rs.getLong("participant_id");
                participantIds.add(participantId);
            }
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }
        List<Participant> participants = new ArrayList<>();
        String placeholders = String.join(",", Collections.nCopies(participantIds.size(), "?"));
        String query3 = "SELECT * FROM participants WHERE id IN (" + placeholders + ")";
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query3)) {
            logger.info("Finding participants in list!");
            for (int i = 0; i < participantIds.size(); i++) {
                stmt.setLong(i + 1, participantIds.get(i));
            }
            ResultSet rs = stmt.executeQuery();
            while (rs.next()) {
                Long participantId = rs.getLong("id");
                String participantName = rs.getString("name");
                String participantIdNumber = rs.getString("id_number");
                Long participantEngineSize = rs.getLong("engine_size");
                participants.add(new Participant(participantId, participantName, participantIdNumber, participantEngineSize));
            }
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }
        Race race = new Race(raceId, name, engineSize, date, participants);
        return Optional.of(race);
    }

    @Override
    public List<Race> findAll() {
        String query = "SELECT * FROM races";
        List<Race> races = new ArrayList<>();
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query)) {
            logger.info("Finding all races!");
            ResultSet rs = stmt.executeQuery();
            while (rs.next()) {
                Long raceId = rs.getLong("id");
                findById(raceId).ifPresent(races::add);
            }
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }
        return races;
    }

    @Override
    public void signUpParticipant(Participant participant, Race race) {
        String query = "INSERT INTO races_participants (race_id, participant_id) VALUES (?, ?)";
        try (Connection conn = DriverManager.getConnection(url);
             PreparedStatement stmt = conn.prepareStatement(query)) {
            logger.info("Signing up participant: {}", participant);
            stmt.setLong(1, race.getId());
            stmt.setLong(2, participant.getId());
            stmt.executeUpdate();
        } catch (SQLException e) {
            logger.error("Error connecting: {}", e.getMessage(), e);
        }
    }
}

File: TemaProiect/motoProjectC#/motoProjectCSharp/motoProjectCSharp/repos/RaceRepo.cs
namespace motoProjectCSharp.repos;

using System.Collections.Generic;
using motoProjectCSharp.domain;
using Microsoft.Data.SqlClient;
using System.Linq;
using log4net;


public class RaceRepo : IRaceRepo
    {
        private readonly string _url;
        private static readonly ILog Logger = LogManager.GetLogger(typeof(RaceRepo));

        public RaceRepo(string url)
        {
            _url = url;
        }

        public void Save(Race race)
        {
            string query = "INSERT INTO races (name, engine_size, date) VALUES (@name, @engine_size, @date)";
            try
            {
                using (var conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (var cmd = new SqlCommand(query, conn))
                    {
                        Logger.Info($"Saving race: {race}");
                        cmd.Parameters.AddWithValue("@name", race.Name);
                        cmd.Parameters.AddWithValue("@engine_size", race.EngineSize);
                        cmd.Parameters.AddWithValue("@date", race.Date);
                        cmd.ExecuteNonQuery();
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error($"Error saving race: {e.Message}", e);
            }
        }

        public Race? FindById(long id)
        {
            string query1 = "SELECT * FROM races WHERE id = @id";
            string query2 = "SELECT * FROM races_participants WHERE race_id = @race_id";

            long raceId = 0, engineSize = 0;
            string? name = null, date = null;
            try
            {
                using (var conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (var cmd = new SqlCommand(query1, conn))
                    {
                        Logger.Info($"Finding race by id: {id}");
                        cmd.Parameters.AddWithValue("@id", id);
                        using (var reader = cmd.ExecuteReader())
                        {
                            if (reader.Read())
                            {
                                raceId = reader.GetInt64(reader.GetOrdinal("id"));
                                name = reader.GetString(reader.GetOrdinal("name"));
                                engineSize = reader.GetInt64(reader.GetOrdinal("engine_size"));
                                date = reader.GetString(reader.GetOrdinal("date"));
                            }
                        }
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error($"Error finding race by id: {e.Message}", e);
            }

            if (raceId == 0)
            {
                Logger.Info($"Race not found: {id}");
                return null;
            }

            List<long> participantIds = new List<long>();
            try
            {
                using (var conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (var cmd = new SqlCommand(query2, conn))
                    {
                        Logger.Info($"Finding participants by race id: {id}");
                        cmd.Parameters.AddWithValue("@race_id", id);
                        using (var reader = cmd.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                participantIds.Add(reader.GetInt64(reader.GetOrdinal("participant_id")));
                            }
                        }
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error($"Error finding participants by race id: {e.Message}", e);
            }

            List<Participant> participants = new List<Participant>();
            string placeholders = string.Join(",", participantIds.Select((participantId, index) => "@id" + index));
            string query3 = "SELECT * FROM participants WHERE id IN (" + placeholders + ")";
            try
            {
                using (var conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (var cmd = new SqlCommand(query3, conn))
                    {
                        Logger.Info("Finding participants in list!");
                        for (int i = 0; i < participantIds.Count; i++)
                        {
                            cmd.Parameters.AddWithValue("@id" + i, participantIds[i]);
                        }
                        using (var reader = cmd.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                var participantId = reader.GetInt64(reader.GetOrdinal("id"));
                                var participantName = reader.GetString(reader.GetOrdinal("name"));
                                var participantIdNumber = reader.GetString(reader.GetOrdinal("id_number"));
                                var participantEngineSize = reader.GetInt64(reader.GetOrdinal("engine_size"));
                                participants.Add(new Participant(participantId, participantName, participantIdNumber, participantEngineSize));
                            }
                        }
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error($"Error finding participants: {e.Message}", e);
            }

            return new Race(raceId, name ?? "default", engineSize, date ?? "default", participants);
        }

        public List<Race> FindAll()
        {
            string query = "SELECT * FROM races";
            List<Race> races = new List<Race>();
            try
            {
                using (var conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (var cmd = new SqlCommand(query, conn))
                    {
                        Logger.Info("Finding all races!");
                        using (var reader = cmd.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                var raceId = reader.GetInt64(reader.GetOrdinal("id"));
                                var race = FindById(raceId);
                                if (race != null)
                                {
                                    races.Add(race);
                                }
                            }
                        }
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error($"Error finding races: {e.Message}", e);
            }
            return races;
        }

        public void SignUpParticipant(Participant participant, Race race)
        {
            string query = "INSERT INTO races_participants (race_id, participant_id) VALUES (@race_id, @participant_id)";
            try
            {
                using (var conn = new SqlConnection(_url))
                {
                    conn.Open();
                    using (var cmd = new SqlCommand(query, conn))
                    {
                        Logger.Info($"Signing up participant: {participant}");
                        cmd.Parameters.AddWithValue("@race_id", race.Id);
                        cmd.Parameters.AddWithValue("@participant_id", participant.Id);
                        cmd.ExecuteNonQuery();
                    }
                }
            }
            catch (SqlException e)
            {
                Logger.Error($"Error signing up participant: {e.Message}", e);
            }
        }
    }